### PR TITLE
Cm/react testing

### PR DESCRIPTION
**Pull request by CheezItMan**
## Description
Apologies for the delay.  I'd written a lot yesterday and decided I didn't like it, so spent this morning rewriting.  

## Questions for the Author
- [ ] Have you updated the weekly learning goals in the pedagogy resource
- [x] Have you added links to the source (lucidchart, draw.io) for any diagram

## Feedback Requested
- Is this too much content?
- Is the use of Enzyme clear enough?
- Is this useful enough for students at this point in the curriculum

### CODE REVIEW

File: React/README.md [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
+1.  [`props`](props.md)
+1.  [Styling](styling.md)
+1.  [`state`](state.md)
+1.  [Event Handling](events.md)
+1.  [Creating Components Part 2](creating-components-2.md)
+1.  [Forms](forms.md)
+  - [Exercise: controlled forms](exercises/forms-worksheet.md)
+1.  [Styling](styling.md)
+1.  [Prop-Types](prop-types.md)
+1.  [Testing](testing.md)
```
Comment (kariabancroft) [inline React/README.md]: Should we just review this resource and not the "react-testing.md" one?

Comment (CheezItMan) [reply React/README.md]: Fixed, apologies

File: React/testing.md [hunk 2]
```diff
@@ -1 +1,299 @@
...
+
+```javascript
+// src/NewPetForm.test.js
+import React from 'react';
+import NewPetForm from './NewPetForm';
+import { mount } from 'enzyme';
+
+describe('NewPetForm', () => {
+  test('that it matches an existing snapshot', () => {
+    const wrapper = mount( <NewPetForm addPetCallback={() => {} } />);
```
Comment (kariabancroft) [inline React/testing.md]: I'd love to see a line-by-line explanation of what all of this code is doing  in the test block - maybe comments on each line?

File: React/testing.md [hunk 3]
```diff
@@ -1 +1,299 @@
...
+Notice that the snapshot file looks a lot like the JSX rendered by the component, but it's a generated JSON file making it easier to compare to new snapshots.  You generally won't need to look at these files directly, but looking at the snapshot file should help you understand how Jest compares the current snapshot to the rendered content from the test.
+
+
+### Creating a Shallow Snapshot with enzyme
+
+This type of snapshot can work just fine for small components without subcomponents.  However with a complicated container component with dozens of subcomponents the snapshot can be thousands of lines long.  It would also be very fragile.  Any time one subcomponent changed, the snapshot would fail with an enormous diff.  A test that fails too often and is hard to parse gets ignored becoming useless.  **Snapshot test files should as a rule be as small as practical and specific to the component being tested**.
+
+To keep snapshot files smaller we can engage in **shallow rendering**.
+
+### Shallow Rendering
```
Comment (kariabancroft) [inline React/testing.md]: I like that you do the regular `mount` and the `shallow` to show the difference. Can we just run the big snapshot first before discussing why we'd want the `shallow` one?

File: React/testing.md [hunk 4]
```diff
@@ -1 +1,299 @@
...
+Use `mount` when you need to test the interaction between a container and child component.  Otherwise for unit testing use `shallow` almost exclusively.  You should **avoid** using `mount` for snapshot testing.
+
+| Enzyme Function | Used for |  Description |
+| ------ | ------ | ----- |
+|  mount  |  Deep Rendering | Renders the entire component and subcomponent to test their interactions  |
+|  shallow  |  shallow rendering  | Used for tests on a single component in isolation.
+
+## Testing User Interaction
+
+With snapshot testing we have tested how React components **render** to the DOM.  Next we need to test how the componets respond to user interaction.
```
Comment (kariabancroft) [inline React/testing.md]: components typo

File: React/testing.md [hunk 5]
```diff
@@ -1 +1,299 @@
...
+      value: 'Bob',
+    },
+  });
+  // Force the component to update (setState is async)
+  wrapper.update();
+  // Find the updated component
+  nameField = wrapper.find('input[name="name"]');
+
+  // Assert
+  expect(nameField.getElement().props.value).toEqual('Bob');
```
Comment (kariabancroft) [inline React/testing.md]: Is this `props` referring literally to the HTML properties? This would be good to clarify since `props` is really loaded

File: React/package.json [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (kariabancroft) [inline React/package.json]: Do you need this file in the textbook?

File: React/testing.md [hunk 7]
```diff
@@ -1 +1,299 @@
...
+## Learning Goals
+By the end of this lesson you should be able to:
+- Explain the terms _Snapshot Test_, and _shallow_ and _deep_ rendering
+- Write tests to verify components exist and actions perform as expected
+
+## Getting Started
+
+In this lesson we will be learning testing through the React pets project we worked on earlier.
+
+You can clone the repository [here](https://github.com/AdaGold/react-pets), and switch to the `solution` branch, if needed.
```
[the remark this answers is not in the document]
Comment (CheezItMan) [reply React/testing.md]: Not really as I'm not testing the API material.  I wanted this to be API independent so we could switch the order later on.

File: React/testing.md [hunk 8]
```diff
@@ -1 +1,299 @@
...
+
+When a component changes, we can be alerted and make corrections, or update the snapshot for future test runs.
+
+**Question:** Can you do TDD with a Snapshot test?
+
+### Creating our First Snapshot
+
+Lets create a snapshot test for one of our untested components, `NewPetForm`.
+
+First create `src/NewPetForm.test.js`
```
[the remark this answers is not in the document]
Comment (CheezItMan) [reply React/testing.md]: There are two schools of thought, one group wants to create a `__tests__` folder and put tests in there.  The other is just to name the components `COMPONENTNAME.test.js`.  

The nice thing about keeping the test in the same folder with the component is that it's slightly easier to move components between projects (just selecting the component js file, css file and test file together).
